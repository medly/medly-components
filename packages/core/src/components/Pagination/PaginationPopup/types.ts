export interface PaginationPopupProps {
    prevPageNumber: number;
    nextPageNumber: number;
    onClickHandler: (page: number) => void;
}
