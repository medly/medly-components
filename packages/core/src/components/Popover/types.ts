import { HTMLProps } from '@medly-components/utils';
import Popup from './Popup';

export type Placement =
    | 'top-start'
    | 'top'
    | 'top-end'
    | 'right-start'
    | 'right'
    | 'right-end'
    | 'bottom-end'
    | 'bottom'
    | 'bottom-start'
    | 'left-end'
    | 'left'
    | 'left-start';

export type InteractionType = 'click' | 'hover';

export type PopoverContextType = [
    boolean /** Current open state */,
    React.Dispatch<React.SetStateAction<boolean>> /** Function to be called to toggle the open state */
];

export type StaticProps = {
    Popup: typeof Popup;
    Context: React.Context<PopoverContextType>;
};

export interface PopoverProps extends HTMLProps<HTMLDivElement> {
    /** Interaction type */
    interactionType?: InteractionType;
}
