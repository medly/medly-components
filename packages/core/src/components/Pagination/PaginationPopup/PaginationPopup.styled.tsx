import { centerAligned } from '@medly-components/utils';
import styled from 'styled-components';
import Text from '../../Text';
import { getPageNumberButtonStyleByState, PageNumberButton } from '../Pagination.styled';

export const PaginationBackgroundStyled = styled('div')`
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: transparent;
    z-index: 4;
    ${centerAligned()}
`;

export const PageEllipsisOverlay = styled('div')`
    max-height: 20rem;
    display: flex;
    align-items: center;
    flex-direction: column;
    overflow: auto;
    background: ${({ theme }) => theme.colors.white};
    box-shadow: 0 0.2rem 0.8rem ${({ theme }) => theme.colors.grey[400]};
    border-radius: 0.4rem;

    & > ${PageNumberButton} {
        display: block;
        border-radius: 0;
        margin: 0;
        height: 4rem;
        width: 4rem;
        padding: 0.4rem;

        & > ${Text.Style} {
            line-height: 3.2rem;
        }

        &:hover {
            ${getPageNumberButtonStyleByState('overlayPageNumber', 'hovered')}
        }

        &:active {
            ${getPageNumberButtonStyleByState('overlayPageNumber', 'hovered')}
        }
    }
`;
