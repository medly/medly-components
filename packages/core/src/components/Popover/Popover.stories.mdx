import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import { Popover } from './Popover.tsx';
import { withKnobs, boolean, select, color } from '@storybook/addon-knobs';
import {
    DummyDiv,
    DummyWrapper,
    DummyPopover,
    interactionTypes,
    PlacementContent,
    PlacementPreviewWrapper,
    PlacementPreviewColumn,
    placements
} from './Popover.stories.tsx';

<Meta title="Core" component={Popover} />

# Popover

A simple popover component which will show a popup on hovering/clicking on the content. The basic structure would be like below:

```tsx
<Popover interactionType="hover">
    <YourContent>Hello</YourContent>
    <Popover.Popup placement="bottom">
        <YourPopoverContent>This is Popover</YourPopoverContent>
    </Popover.Popup>
</Popover>
```

<Preview withToolbar>
    <Story name="Popover" decorators={[withKnobs]}>
        <DummyWrapper>
            <Popover interactionType={select('Interaction Type', interactionTypes, 'hover')}>
                <DummyDiv>Hello</DummyDiv>
                <Popover.Popup
                    placement={select('Placement', placements, 'right')}
                    fullWidth={boolean('Full Width', false)}
                    fullHeight={boolean('Full Height', false)}
                    withArrow={boolean('With Arrow', true)}
                    bgColor={color('Background color', 'white')}
                >
                    <DummyPopover>This is a Popover</DummyPopover>
                </Popover.Popup>
            </Popover>
        </DummyWrapper>
    </Story>
</Preview>

### With Arrow

The popup can be used with/without an arrow using the `withArrow` prop. The arrow makes the popover appear like a tooltip and can be adjusted to various positions. You can also change the background color using `bgColor` prop.

### Without Arrow
<Preview>
    <DummyWrapper>
        <Popover>
            <DummyDiv>Hello</DummyDiv>
            <Popover.Popup placement="right">
                <DummyPopover>This is a Popover</DummyPopover>
            </Popover.Popup>
        </Popover>
    </DummyWrapper>
</Preview>

### Placements

<Preview>
    <PlacementPreviewWrapper>
        {placements.map(placement => (
            <PlacementPreviewColumn>
                <Popover>
                    <PlacementContent>{placement}</PlacementContent>
                    <Popover.Popup
                        placement={placement}
                        withArrow
                    >
                        <Popover>This is a Popover</Popover>
                    </Popover.Popup>
                </Popover>
            </PlacementPreviewColumn>
        ))}
    </PlacementPreviewWrapper>
</Preview>

#### Note

> You can also override the style of `Popover.Style` & `Popover.Popup.Style`

### Accessing current state

We have also exposed `Popover.Context`, so that you can find out the latest open state. Popover context can be used like below:

```tsx
const [isOpen, setOpenState] = useContext(Popover.Context);
```

Type of Popover Context is:

```tsx
import * as React from 'react';
export type PopoverContextType = [
    boolean /** Current open state */,
    React.Dispatch<React.SetStateAction<boolean>> /** Function to be called to toggle the open state */
];
```

### Popover Props

<Props of={Popover} />

### Popover.Popup Props

<Props of={Popover.Popup} />
