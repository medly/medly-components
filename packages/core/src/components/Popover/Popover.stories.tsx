import { centerAligned, styled, css } from '@medly-components/utils';
import { InteractionType, Placement } from './types';

export const placements: Placement[] = [
    'top-start',
    'top',
    'top-end',
    'right-start',
    'right',
    'right-end',
    'bottom-start',
    'bottom',
    'bottom-end',
    'left-end',
    'left',
    'left-start'
];

export const interactionTypes: InteractionType[] = ['hover', 'click'];

const commonContentStyles = () => css`
    color: white;
    width: 150px;
    height: 50px;
    background-color: steelblue;
    display: inline-block;
    border-radius: 4px;
    ${centerAligned()};
`;

export const DummyDiv = styled('div')`
    ${commonContentStyles()};
    font-size: 20px;
`;

export const DummyWrapper = styled.div`
    width: 100%;
    height: 100%;
    ${centerAligned()};
`;

export const DummyPopover = styled.div`
    padding: 1rem 2rem;
    border: 1px solid;
    border-radius: 4px;
`;

export const PlacementPreviewWrapper = styled.div`
    display: grid;
    grid-row-gap: 20px;
    grid-template-columns: repeat(3, 1fr);
`;

export const PlacementPreviewColumn = styled.div`
    ${centerAligned()};
`;

export const PlacementContent = styled.div`
    ${commonContentStyles()};
    font-size: 16px;
    text-transform: capitalize;
`;
