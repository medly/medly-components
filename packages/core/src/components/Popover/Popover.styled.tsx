import styled from 'styled-components';
import { PopoverProps } from './types';

export const Wrapper = styled('div')<PopoverProps>`
    position: relative;
    display: inline-flex;
`;
