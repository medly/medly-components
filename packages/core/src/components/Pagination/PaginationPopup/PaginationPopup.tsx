import { memo, useContext } from 'react';
import Popover from '../../Popover';
import Text from '../../Text';
import { addPageItems } from '../helper';
import { PageNumberButton } from '../Pagination.styled';
import { PageEllipsisOverlay, PaginationBackgroundStyled } from './PaginationPopup.styled';
import { PaginationPopupProps } from './types';
import type { FC } from 'react';

export const PaginationPopup: FC<PaginationPopupProps> = memo(({ prevPageNumber, nextPageNumber, onClickHandler }) => {
    const [isPopoverVisible, setPopoverVisibility] = useContext(Popover.Context),
        onClickHandlerWrapper = (pageNumber: any) => () => {
            setPopoverVisibility(false);
            onClickHandler(pageNumber);
        };
    return (
        <>
            <PageNumberButton key={prevPageNumber + '...'}>
                <Text textVariant="h5" textAlign="center" textWeight="Medium">
                    ...
                </Text>
            </PageNumberButton>
            {isPopoverVisible && <PaginationBackgroundStyled />}
            <Popover.Popup placement="top">
                <PageEllipsisOverlay>
                    {addPageItems(prevPageNumber + 1, nextPageNumber - 1, []).map(item => (
                        <PageNumberButton key={item} onClick={onClickHandlerWrapper(item)}>
                            <Text textVariant="h5" textAlign="center" textWeight="Medium">
                                {item}
                            </Text>
                        </PageNumberButton>
                    ))}
                </PageEllipsisOverlay>
            </Popover.Popup>
        </>
    );
});
PaginationPopup.displayName = 'PaginationPopup';
