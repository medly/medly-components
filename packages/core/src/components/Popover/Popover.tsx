import { useOuterClickNotifier, WithStyle } from '@medly-components/utils';
import React, { FC, useCallback, useRef, useState } from 'react';
import { PopoverContext } from './Popover.context';
import { Wrapper } from './Popover.styled';
import Popup from './Popup';
import { PopoverProps, StaticProps } from './types';

const Component: FC<PopoverProps> = React.memo(({ interactionType, ...restProps }) => {
    const openState = useState(false),
        [, setOpenState] = openState,
        wrapperRef = useRef(null);

    const handleMouseOver = useCallback(() => interactionType === 'hover' && setOpenState(true), [interactionType]),
        handleMouseLeave = useCallback(() => interactionType === 'hover' && setOpenState(false), [interactionType]),
        handleMouseClick = useCallback(() => interactionType === 'click' && setOpenState(val => !val), [interactionType]);

    useOuterClickNotifier(() => interactionType === 'click' && setOpenState(false), wrapperRef);

    return (
        <PopoverContext.Provider value={openState}>
            <Wrapper
                {...restProps}
                ref={wrapperRef}
                onMouseOver={handleMouseOver}
                onMouseLeave={handleMouseLeave}
                onClick={handleMouseClick}
            />
        </PopoverContext.Provider>
    );
});

Component.displayName = 'Popover';
Component.defaultProps = {
    id: 'medly-popover-wrapper',
    interactionType: 'hover'
};
export const Popover: FC<PopoverProps> & WithStyle & StaticProps = Object.assign(Component, {
    Popup,
    Style: Wrapper,
    Context: PopoverContext
});
