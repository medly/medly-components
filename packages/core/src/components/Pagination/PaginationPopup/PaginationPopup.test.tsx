import { cleanup, fireEvent, render, screen } from '@test-utils';
import PaginationPopup from '.';
import Popover from '../../Popover';

describe('Pagination popup component', () => {
    const renderer = (props: any = { prevPageNumber: 1, nextPageNumber: 3 }) => {
        const mockOnPageClick = jest.fn();
        return {
            mockOnPageClick,
            ...render(
                <Popover interactionType="click">
                    <PaginationPopup {...props} onClickHandler={mockOnPageClick} />
                </Popover>
            )
        };
    };

    afterEach(cleanup);

    it('should render correctly with all the default props', () => {
        const { container } = renderer();
        expect(container).toMatchSnapshot();
    });

    it('should call onClick handler with expected page when selected from ellipsis popover', () => {
        const { mockOnPageClick } = renderer();
        fireEvent.click(screen.getByRole('button', { name: '...' }));
        fireEvent.click(screen.getByRole('button', { name: '2' }));
        expect(mockOnPageClick).toBeCalledWith(2);
    });
});
