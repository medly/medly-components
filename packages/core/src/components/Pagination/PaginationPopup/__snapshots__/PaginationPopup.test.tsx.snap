// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination popup component should render correctly with all the default props 1`] = `
.c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c3 {
  margin: 0;
  color: inherit;
  font-size: 1.2rem;
  font-weight: 600;
  -webkit-letter-spacing: 0.12rem;
  -moz-letter-spacing: 0.12rem;
  -ms-letter-spacing: 0.12rem;
  letter-spacing: 0.12rem;
  line-height: 2.0rem;
  text-align: center;
}

.c1 {
  margin: 0.4rem;
  height: 3.2rem;
  width: 3.2rem;
  display: block;
  border-radius: 50%;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: none;
  padding: 0;
  text-align: center;
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  color: #435465;
  background: #ffffff;
  color: #435465;
  background: #ffffff;
  border-radius: 50%;
}

.c1:hover {
  cursor: pointer;
}

.c1:focus {
  outline: none;
}

.c1:disabled {
  cursor: not-allowed;
}

.c1:hover {
  color: #303C48;
  background: #F7F8F9;
}

.c1:active {
  font-weight: 700;
  color: #13181D;
  background: #eff2f4;
}

.c1 .c2 {
  font-size: 1.2rem;
  font-weight: 700;
  -webkit-letter-spacing: 0.12rem;
  -moz-letter-spacing: 0.12rem;
  -ms-letter-spacing: 0.12rem;
  letter-spacing: 0.12rem;
  line-height: 2.0rem;
  font-family: Open Sans,Helvetica Neue,Helvetica,Arial,sans-serif;
  font-weight: 600;
}

<div>
  <div
    class="c0"
    id="medly-popover-wrapper"
  >
    <button
      class="sc-AxheI c1"
      type="button"
    >
      <h5
        class="c2 c3"
      >
        ...
      </h5>
    </button>
  </div>
</div>
`;
